connectd/connectd.c

```diff
@@ -1898,6 +1900,141 @@ static void dev_suppress_gossip(struct daemon *daemon, const u8 *msg)
+	case AF_UNIX:
+		if (addrlen == sizeof(un->sun_family))
+			return tal_fmt(ctx, "unix socket <unnamed>");
+		else
+			return tal_fmt(ctx, "unix socket %s", un->sun_path);
+	case AF_INET:
+		if (!inet_ntop(sa->ss_family, &in->sin_addr, addr, sizeof(addr)))
+			return tal_fmt(ctx, "IPv4 socket <badaddr>");
+		else
+			return tal_fmt(ctx, "IPv4 socket %s:%u",
+				       addr, ntohs(in->sin_port));
+	case AF_INET6:
+		if (!inet_ntop(sa->ss_family, &in6->sin6_addr, addr, sizeof(addr)))
+			return tal_fmt(ctx, "IPv6 socket <badaddr>");
+		else
+			return tal_fmt(ctx, "IPv6 socket %s:%u",
+				       addr, ntohs(in6->sin6_port));
+	default:
+		return tal_fmt(ctx, "unknown family %u (**BROKEN**)",
+			       (unsigned)sa->ss_family);
+	}
+}
+
+static void describe_fd(int fd)
+{
+	struct sockaddr_storage sa;
+	socklen_t addrlen = sizeof(sa);
+
+	if (getsockname(fd, (void *)&sa, &addrlen) != 0) {
+		status_broken("dev_report_fds: %i cannot get sockname (%s)",
+			      fd, strerror(errno));
+		return;
+	}
+	status_info("dev_report_fds: %i name %s", fd, addr2name(tmpctx, &sa, addrlen));
+
+	if (getpeername(fd, (void *)&sa, &addrlen) != 0)
+		return;
+	status_info("dev_report_fds: %i peer %s", fd, addr2name(tmpctx, &sa, addrlen));
+}
+
+static const char *io_plan_status_str(enum io_plan_status status)
+{
+	switch (status) {
+	case IO_UNSET: return "IO_UNSET";
+	case IO_POLLING_NOTSTARTED: return "IO_POLLING_NOTSTARTED";
+	case IO_POLLING_STARTED: return "IO_POLLING_STARTED";
+	case IO_WAITING: return "IO_WAITING";
+	case IO_ALWAYS: return "IO_ALWAYS";
+	}
+	return "INVALID-STATUS";
+}
+
+/* Stupid and slow, but machines are fast! */
+static const tal_t *find_tal_ptr(const tal_t *root, const tal_t *p)
+{
+	if (root == p)
+		return root;
+
+	for (tal_t *t = tal_first(root); t; t = tal_next(t)) {
+		const tal_t *ret = find_tal_ptr(t, p);
+		if (ret)
+			return ret;
+	}
+	return NULL;
+}
+
+/* Looks up ptr in hash tree, to try to find name */
+static const char *try_tal_name(const tal_t *ctx, const void *p)
+{
+	const tal_t *t = find_tal_ptr(NULL, p);
+	if (t)
+		return tal_name(t);
+	return tal_fmt(ctx, "%p", p);
+}
+
+static void dev_report_fds(struct daemon *daemon, const u8 *msg)
+{
+	for (int fd = 3; fd < 4096; fd++) {
+		bool listener;
+		const struct io_conn *c;
+		const struct io_listener *l;
+		if (!isatty(fd) && errno == EBADF)
+			continue;
+		if (fd == HSM_FD) {
+			status_info("dev_report_fds: %i -> hsm fd", fd);
+			continue;
+		}
+		if (fd == GOSSIPCTL_FD) {
+			status_info("dev_report_fds: %i -> gossipd fd", fd);
+			continue;
+		}
+#if DEVELOPER
+		if (fd == daemon->dev_disconnect_fd) {
+			status_info("dev_report_fds: %i -> dev_disconnect_fd", fd);
+			continue;
+		}
+#endif
+		if (fd == daemon->gossip_store_fd) {
+			status_info("dev_report_fds: %i -> gossip_store", fd);
+			continue;
+		}
+		c = io_have_fd(fd, &listener);
+		if (!c) {
+			status_broken("dev_report_fds: %i open but unowned?", fd);
+			continue;
+		} else if (listener) {
```

Comment:
Looks like the CI fails here with the following log 

```
--------------------------- Captured stdout teardown ---------------------------
lightningd-1 2023-04-07T05:56:53.707Z UNUSUAL lightningd: JSON-RPC shutdown
lightningd-1 2023-04-07T05:56:53.738Z DEBUG   lightningd: io_break: start_json_stream
lightningd-1 2023-04-07T05:56:53.739Z DEBUG   lightningd: io_loop_with_timers: main
lightningd-1 2023-04-07T05:56:53.804Z INFO    connectd: dev_report_fds: 3 -> hsm fd
lightningd-1 2023-04-07T05:56:53.806Z INFO    connectd: dev_report_fds: 4 -> gossipd fd
lightningd-1 2023-04-07T05:56:53.808Z **BROKEN** connectd: dev_report_fds: 5 open but unowned?
lightningd-1 2023-04-07T05:56:53.810Z INFO    connectd: dev_report_fds: 6 -> listener (0x113c08 (backtrace disabled))
lightningd-1 2023-04-07T05:56:53.811Z INFO    connectd: dev_report_fds: 6 name IPv4 socket 127.0.0.1:37697
lightningd-1 2023-04-07T05:56:53.818Z INFO    connectd: dev_report_fds: 7 -> IN=IO_POLLING_NOTSTARTED:0x1f2168 (backtrace disabled)+0x11f1fa (backtrace disabled)(connectd/connectd.c:239:struct peer), OUT=IO_WAITING:(nil) (backtrace disabled)+0x11e887 (backtrace disabled)(connectd/connectd.c:239:struct peer)
lightningd-1 2023-04-07T05:56:53.819Z INFO    connectd: dev_report_fds: 7 name IPv4 socket 127.0.0.1:55414
lightningd-1 2023-04-07T05:56:53.821Z INFO    connectd: dev_report_fds: 7 peer IPv4 socket 127.0.0.1:42409
lightningd-1 2023-04-07T05:56:53.822Z INFO    connectd: dev_report_fds: 8 -> gossip_store
lightningd-1 2023-04-07T05:56:53.822Z INFO    connectd: dev_report_fds: 9 -> IN=IO_POLLING_NOTSTARTED:0x1504ce (backtrace disabled)+0x11ea3c (backtrace disabled)(connectd/multiplex.c:1064:struct subd), OUT=IO_WAITING:(nil) (backtrace disabled)+0x11eb0f (backtrace disabled)(connectd/multiplex.c:1064:struct subd)
lightningd-1 2023-04-07T05:56:53.823Z INFO    connectd: dev_report_fds: 9 name unix socket <unnamed>
lightningd-1 2023-04-07T05:56:53.824Z INFO    connectd: dev_report_fds: 9 peer unix socket <unnamed>
lightningd-1 2023-04-07T05:56:53.824Z DEBUG   connectd: REPLY WIRE_CONNECTD_START_SHUTDOWN_REPLY with 0 fds
lightningd-1 2023-04-07T05:56:53.826Z DEBUG   lightningd: io_break: connectd_start_shutdown_reply
```

Reply:
but this looks like fine? maybe it is just a `status_info`?

Reply:
No that's a real bug, though it may be valgrind, which seems not to always close our O_CLOEXEC fd. I've disabled this check under valgrind for now.